keps/0029-operator-dependencies.md

```diff
@@ -0,0 +1,205 @@
+
+```yaml
+apiVersion: kudo.dev/v1beta1
+name: "kafka"
+operatorVersion: "1.3.1"
+kudoVersion: 0.12.0
+kubernetesVersion: 1.14.8
+appVersion: 2.5.0
+url: https://kafka.apache.org/
+tasks:
+  - name: zookeeper-operator
+    kind: Operator
+    spec:
+        package: zookeeper
+
+  ...
+
+plans:
+  deploy:
+    strategy: serial
+    phases:
+      - name: deploy-zookeeper
+        strategy: serial
+        steps:
+          - name: dummy
+            tasks:
+              - zookeeper-operator
+      - name: deploy-kafka
+        strategy: serial
+        steps:
+          ...
+```
+
+The `zookeeper-operator` task specification is equivalent to `kudo install zookeeper` CLI command which installs the Zookeeper package from the official repo. Here is a complete `Operator` task specification:
+
+```yaml
+tasks:
+- name: demo
+  kind: Operator
+  spec:
+    package: # required, either repo package name, local package folder or an URL to package tarball
+    repo: # optional, name of local repository configuration to use
+    appVersion: # optional, a specific app version in the official repo, defaults to the most recent one
+    operatorVersion: # optional, a specific operator version in the official repo, defaults to the most recent one
+    instanceName: # optional, the instance name
+```
+
+As you can see, this closely mimics the `kudo install` CLI command [options](https://github.com/kudobuilder/kudo/blob/master/pkg/kudoctl/cmd/install.go#L56) because at the end the later will be executed to install the operator. We omit `parameters` and `parameterFile` options at this point as they are discussed in detail [below](#dependencies-parameterization)
+
+#### Deployment
+
+##### Client-Side
+
+Upon execution of `kudo install ...` command with the above operator definition, CLI will:
+
+1. Collect all operator dependencies by analyzing the `deploy` plan of the top-level operator and compose a list of all operator dependencies (tasks with the kind `Operator`) including **transitive dependencies**
+2. Install all collected operators **skipping the instances** (same as `kudo install ... --skip-instance`). Note that since we do not create instances here we can install them in any order
+3. Proceed with the installation of the top-level operator as usual (create `Operator`, `OperatorVersion` and `Instance` resources)
+
+Since we do this step on the client-side we have access to the full functionality of the `install` command including installing operators from the file system. This will come very handy during the development and debugging which arguably becomes more complex with dependencies.
+
+##### Server-Side
+
+Upon receiving a new operator Instance with dependencies KUDO mangers workflow engine will:
+
+1. Build a [dependency graph](https://en.wikipedia.org/wiki/Dependency_graph) by transitively expanding top-level `deploy` plan using operator-tasks as vertices and their execution order (`a` needs `b` to be installed first) as edges
```

Comment:
So you cannot have operator deps in plans other than deploy? Not sure if this is a problem, but this limitation should probably be mentioned somewhere.
Or perhaps there should be another separate special "dependencies" plan, or this should walk the list of all defined tasks... just wondering..

Reply:
No, the later sections show dependencies for e.g. `upgrade` and `update` plans. I'm not sure if it is meaningful to allow operator-task outside of these plans but theoretically, it is possible 🤷